File: 2023/python/days/01.py
import re
from pathlib import Path

from utils.logger import logger


def main() -> int:
    input_file = Path(__file__).parent / "resources" / "input.01.txt"

    solution = 0
    with input_file.open("r") as f:
        for index, line in enumerate(f):
            first_number = None
            last_number = None

            """matches = re.findall(
                r"\d|one|two|three|four|five|six|seven|eight|nine",
                line,
            )
            if len(matches) == 1:
                first_number = matches.group(0)
                last_number = first_number
            else:"""

            # First number in Line
            for c in line:
                if c.isnumeric():
                    first_number = c
                    break
            # First number in Reversed Line
            for c in line[::-1]:
                if c.isnumeric():
                    last_number = c
                    break
            add = int(f"{first_number}{last_number}")
            logger.debug(f"line: {index+1}, {add=}")
            solution += add
    return solution

File: 2023/go/days/01.go
package advent

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"slices"
	"strconv"
)

type DayOne struct {
}

var wordToNumber = map[string]int{
	"one":   1,
	"two":   2,
	"three": 3,
	"four":  4,
	"five":  5,
	"six":   6,
	"seven": 7,
	"eight": 8,
	"nine":  9,
}

func intify(number string) (int, error) {
	if val, ok := wordToNumber[number]; ok {
		return val, nil
	}
	return strconv.Atoi(number)
}

func MatchDigit(input string, reverse bool) (string, error) {
	line := []rune(input) // Reverse works on the reference, so we need to assign line to variable
	numbers := []rune(`one|two|three|four|five|six|seven|eight|nine`)
	digits := `\d` + `|`

	if reverse {
		slices.Reverse(line)
		slices.Reverse(numbers)
	}
	re := regexp.MustCompile(digits + string(numbers))
	match := re.Find([]byte(string(line)))

	if reverse {
		slices.Reverse(match)
	}
	digit, err := intify(string(match))
	if err != nil {
		return "", fmt.Errorf("cannot convert to integer: %v", err)
	}
	return fmt.Sprintf("%d", digit), nil
}

func (d DayOne) Solve() (string, error) {

	wd, err := os.Getwd()
	if err != nil {
		return "", err
	}
	inputFile := filepath.Join(filepath.Dir(wd), "resources", "input-01.txt")

	f, err := os.Open(inputFile)
	if err != nil {
		return "", err
	}
	scanner := bufio.NewScanner(f)
	sum := 0
	for scanner.Scan() {
		line := scanner.Text()
		first_match, err := MatchDigit(line, false)
		if err != nil {
			return "", err
		}
		last_match, err := MatchDigit(line, true)
		if err != nil {
			return "", err
		}

		add, err := strconv.Atoi(first_match + last_match)
		if err != nil {
			return "", err
		}
		sum += add
	}
	return fmt.Sprintf("%d", sum), nil
}
